space_packet_parser/encodings.py

```diff
@@ -243,104 +262,138 @@ def __init__(self,
-            nbits_left = len(packet.raw_data) - packet.raw_data.pos
-            orig_pos = packet.raw_data.pos
-            string_buffer = packet.raw_data.read_as_bytes(nbits_left - nbits_left % 8)
-            # Reset the original position because we only wanted to look ahead
-            packet.raw_data.pos = orig_pos
-            try:
-                strlen_bits = string_buffer.index(self.termination_character) * 8
-            except ValueError as exc:
-                # Termination character not found in the string buffer
-                raise ValueError(f"Reached the end of the packet data without finding the "
-                                 f"termination character {self.termination_character}") from exc
+                buflen_bits = packet[self.dynamic_length_reference].raw_value
+
+            if self.length_linear_adjuster:
+                buflen_bits = self.length_linear_adjuster(buflen_bits)
         else:
-            raise ValueError("Unable to parse StringParameterType. "
-                             "Didn't contain any way to constrain the length of the string.")
-        if not self.termination_character and self.length_linear_adjuster is not None:
-            # Only adjust if we are not doing this by termination character. Adjusting a length that is objectively
-            # determined via termination character is nonsensical.
-            strlen_bits = self.length_linear_adjuster(strlen_bits)
-        return strlen_bits
-        # pylint: enable=too-many-branches
-
-    def parse_value(self, packet: packets.CCSDSPacket, **kwargs) -> Tuple[str, None]:
-        """Parse a value from packet data, possibly using previously parsed data items to inform parsing.
+            raise ValueError("No raw length specifier found when decoding a string.")
+        return int(buflen_bits)
+
+    def _get_raw_buffer(self, packet: packets.CCSDSPacket) -> bytes:
+        """Get the raw string buffer as bytes. This will include any leading size or termination characters.
+
+        Notes
+        -----
+        If the buffer size is not an integer number of bytes, the bytestring is padded at the end with zeros.
+
+        Parameters
+        ----------
+        packet : packets.CCSDSPacket
+            Packet parsed so far, for referencing previous values
+
+        Returns
+        -------
+        : bytes
+            The raw string buffer, padded on the RHS to a byte boundary if the raw length specified is not an integer
+            number of bytes.
+        """
+        buflen_bits = self._calculate_size(packet)
+        pad_bits = (8 - (buflen_bits % 8)) % 8
+        buflen_bytes = (buflen_bits + pad_bits) // 8
+        if (buflen_bits + pad_bits) % 8 != 0:
+            raise ValueError(f"Error in buffer length math in _get_raw_buffer. "
+                             f"buflen_bits={buflen_bits}, pad_bits={pad_bits}, buflen_bytes={buflen_bytes}")
```

Comment:
On the non-integer number of bytes and padding of the raw string buffer. The only way this happens is if the leading size integer is a weird size. For example, you might have a string buffer size that is maximum 51 bits but the actual string is of dynamic length, specified by a leading size integer that is a uint3. So you actually have a uint3 followed by at most 6 bytes (48 bits) of UTF8 characters. In this case, the raw string buffer is 51 bits because the leading size is important to the string itself and that's what was actually downlinked for that field. I chose to pad it on the RHS so that the first bit in the string is the first bit in the string container.

On the raw/derived value issue. Yeah, this is a bit different semantically because strings aren't "calibrated" but are secondarily parsed. IMO this is sort of a "calibration" in that it is creating a derived value for the string but the raw value of the string is the string buffer that was downlinked in the packet.

Reply:
non-integer number of bytes: OK, now I think I'm following and that makes sense. The encoding/decoding does have an error check below too.

Alternative recommendation: Would it be feasible to rework the logic below in `parse()` with an initial "read the 3 bits" then continue on...

```python
        if self.leading_length_size:
            strlen_bits = packet.raw_data.read_as_int(self.leading_length_size)
            if strlen_bits % 8 != 0:
                raise ValueError(f"String length (in bits) is {strlen_bits}, which is not a multiple of 8. "
                                 "This is an error since strings must be an integer numbers of bytes.")
            return packet.raw_data.read_as_bytes(strlen_bits).decode(self.encoding)

        # continue the other if clauses
        buflen = self._calculate_size(packet)
```

Reply:
Do you mean read the leading size int and use it but not store it in the raw string buffer? I think it's important that the raw buffer include all the information that was encoded. If we read the leading size int and throw it away, IMO it's misleading to a user that is examining the raw values. Padding with zeros at the end is also potentially misleading if the user is looking at string lengths, but I think that's less impactful than losing the leading size information. The raw value we return should have all the information necessary to recreate the derived value.

Reply:
Ahh, that makes sense. That is annoying because I agree that both situations are less than ideal. But if you are trying to debug this, then yes you do need those initial 3 bits telling you the length and it should be in the raw return.

Reply:
Yes, if FSW is doing things cleanly, the leading size should just be a separate field and the string length should be specified using a dynamic reference to that previous parameter. That would totally solve the issue, but XTCE allows this jank...